Improve i18n support

jhsoby opened the pull request:
Avoid use of LEGO messages, so that languages with different
word order from English are free to place numbers and other
things where they need to be in a sentence.

jhsoby committed:
Improve i18n support

Avoid use of LEGO messages, so that languages with different
word order from English are free to place numbers and other
things where they need to be in a sentence.

jhsoby committed:
Improve i18n support

Avoid use of LEGO messages, so that languages with different
word order from English are free to place numbers and other
things where they need to be in a sentence.

jhsoby committed:
Fix missing variable

jhsoby committed:
Remove unused variables

jhsoby committed:
Merge branch 'main' into main

jhsoby commented:
Sorry to make such a huge PR, but I didn't really know how to make it smaller.

Working Netlify deployment here: https://velvety-entremet-ea5c17.netlify.app/

jdlrobson commented:
I realize the LEGO style messages might be challenging, but I fear this might impact the visual display.

I wanted to check it out, but I'm getting the following error?:
```
streak.js:14 Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'longestStreak')
    at r (streak.js:14:33)
    at f (index.js:20:27)
    at App.vue:277:18
r @ streak.js:14
f @ index.js:20
(anonymous) @ App.vue:277
Promise.then
start @ App.vue:272
n.<computed>.n.<computed> @ runtime-dom.esm-bundler.js:1497
f @ runtime-core.esm-bundler.js:195
m @ runtime-core.esm-bundler.js:202
n @ runtime-dom.esm-bundler.js:696Understand this errorAI
```

jhsoby commented:
> I realize the LEGO style messages might be challenging, but I fear this might impact the visual display.

I did my best to keep everything visually the same. The only difference is that the qualifier is no longer on a line of its own, as extracting it from the rest of a sentence is not really possible i18n-wise.

You might be thinking that we could use a message like "n-edits" with English content "$1 edits", and insert that message into the full sentences instead of the number. The problem with that is that "$1 edits" might not work as part of a sentence in other languages. Even in Norwegian, which is closely related to English, we can use "$1 redigeringer" or "$1 redigeringene" as a translation for "$1 edits", depending on the context. And in Swahili, "$1 edits" would be "mabadiliko $1", so not even the internal order there is guaranteed to be the same.

There are a couple of other differences in the final "share" card too, but those are really just changes to the messages (in en.json) that I made to illustrate (in English) how it will look for different languages if they have things in a different order. We can easily change those back to what they used to be for English.

> 
> I wanted to check it out, but I'm getting the following error?:

How/where are you seeing this error? All my testing so far (of the submitted code) has worked fine both in [my Netlify deployment](https://velvety-entremet-ea5c17.netlify.app/) (as well as locally) and in the "[Deployment Preview](https://deploy-preview-6--wikipediayir.netlify.app/)" from @netlify's comment above. I've been testing primarily with my own username in various wikis – maybe you tested it with a username that gets some other results that aren't accounted for?

jhsoby commented:
> How/where are you seeing this error? All my testing so far (of the submitted code) has worked fine both in [my Netlify deployment](https://velvety-entremet-ea5c17.netlify.app/) (as well as locally) and in the "[Deployment Preview](https://deploy-preview-6--wikipediayir.netlify.app/)" from @netlify's comment above. I've been testing primarily with my own username in various wikis – maybe you tested it with a username that gets some other results that aren't accounted for?

Actually, I'm seeing some errors too, but I see them both in current production and in the two Netlify deployments I linked. But I think there are two separate issues:
1. I get a streak-related error in all 4 versions (prod, 2 netlify test deployments, **and** localhost) with the combo Jdlrobson, en.wikipedia.org, 2023. I believe this may have to do with the file `/src/assets/data/shortcut/2023/en.wikipedia.org/Jdlrobson.summary.json`, where all the entries are missing a "timestamp" entry.
2. I get [another error](https://phabricator.wikimedia.org/P71449) in all the production versions (but *not* in my localhost with every username/project combo I've tried, and any year below 2023 (so 2023 works, except for Jdlrobson/en.wikipedia, but 2022 and down don't work for any username/project combo).

jdlrobson commented:
Okay fixed the bug, yeh unrelated.

Regarding the translation of text with numbers I'd like to dig into this more, one bit of UI at a time

# Stats
 
I think it would be good to focus on different parts of the UI, perhaps we could add some Norwegian translations so I can understand this a better?
Perhaps beginning with this part of the interface?:
<img width="301" alt="Screenshot 2024-11-30 at 9 00 24 AM" src="https://github.com/user-attachments/assets/f9c06323-2208-4fc9-93bb-15d633a4181f">

My thinking was the following:
* In the above screenshot, I'd like to treat them like units and avoid any text above the number. I don't think the translation needs to be a word for word translation, if that doesn't work. If this doesn't work in certain languages it could be something like:
```
6
number of times you edited
```

Note:
* numbers are always plural (if there is only 0 or 1 edit I don't think we should be displaying these at all)
* we could pass numbers to each of the messages to aid with  "$1 redigeringer" or "$1 redigeringene" 

Is this particular part of the UI troublesome for Norwegian or can retain the number above the text?

# LEGO approach

>  And in Swahili, "$1 edits" would be "mabadiliko $1", so not even the internal order there is guaranteed to be the same.

<img width="338" alt="Screenshot 2024-11-30 at 9 08 03 AM" src="https://github.com/user-attachments/assets/86792c87-167e-4722-ac29-83509c00b718">

In this situation could you repurpose paragraph-1 and paragraph-3 to consistently create the correct order? That was my intention, but if that doesn't work universally even with number passed in, I'll have to rethink my naivety :) 

e.g. = mabadiliko in 

| Key | Swahili    | English |
| -------- | -------  | ------- |
| paragraph-1   | mabadiliko   |you made    |
| paragraph-3 | (translate of you made)    |edits   |

jhsoby commented:
# Stats
> I think it would be good to focus on different parts of the UI, perhaps we could add some Norwegian translations so I can understand this a better? 

Sure! It was the Norwegian translations I made in Translatewiki that brought me to this repo to try to help make some improvements. :-) So they should show up when @translatewiki makes its first commits.

![Screenshot_20241130_183557](https://github.com/user-attachments/assets/01e783b3-8a38-4390-9a68-5d2b8a83f425)

Here's a screenshot with some Norwegian translations I made in the browser dev tools just to illustrate things.

Some notes:
* **Concisely** distinguishing between the two different thanks is really difficult. I'll try to explain why.
  * Thanks received: The translation I have there is the most concise I could come up with. Back-translated literally, it means "{number} thanks received". That's pretty clear.
  * Thanks given: This is where the problems come in. I could write it in a number of ways with just the number at the top and then some text, but they're all ambiguous and easily confused with the number of thanks received. One alternative would be (back-translated) "{number} users thanked", but that can easily be (mis)understood as "{number} users thanked [you]". So the one I have in the screenshot is "thanked {number} other users".
* We don't really have a word for "streak" in this sense in Norwegian at all, so it's really hard to come up with a concise translation for it (if you have unlimited space and words, everything can be translated – it's trying to stay concise that makes things complicated). But that's not specific to this repo though.
  * The translation in the screenshot now says (back-translated) "{number} longest row". To make it clearer to the users what it actually means, I would like to say "På det meste redigerte du {number} dager på rad", which (back-translated) means "On the peak, you edited {number} days in a row". But that's just too long, at least for the share page.

# Plural
> My thinking was the following:
> 
>     * In the above screenshot, I'd like to treat them like units and avoid any text above the number. I don't think the translation needs to be a word for word translation, if that doesn't work. If this doesn't work in certain languages it could be something like:
> 
> 
> ```
> 6
> number of times you edited
> ```
> 
> Note:
> 
>     * numbers are always plural (if there is only 0 or 1 edit I don't think we should be displaying these at all)

In many languages, especially Slavic and Uralic languages, you don't just have *one* plural, you have several. In Russian, for example, the word for "edits" will be different depending on the preceding number: 20 правок, 21 правка, 22 правки. https://www.mediawiki.org/wiki/PLURAL explains it better than I can. :-)

But I think PLURAL support will be a problem for another day, hehe.

# Lego
> 
>     * we could pass numbers to each of the messages to aid with  "$1 redigeringer" or "$1 redigeringene"
> 
Actually, I just thought of an approach that lets us have the best of both worlds: Both a separate line for the quantity, and keeping the messages still avoid using LEGO messages.

For illustrations, let's use the message named `you-made-impact` in my patch as an example (it's normally the first message you encounter after entering your project and username in the app).

The message currently says:
```json
"you-made-impact": "You made $1 edits across the project",
```
In my patch, the $1 gets replaced with `<strong>234</strong>` if you have made 234 edits, so the result becomes:
```lang=html
You made <strong>234</strong> edits across the project
```
And then CSS makes sure that the `strong` element comes on a separate line.


Now, with the approach I just got the idea for, we could instead do this:

```json
"you-made-impact": "Ỳou made <strong>$1 edits</strong> across the project",
```
If we now replace $1 with something like `<span class="number">234</span>`, we could use CSS to make sure that both the `strong` element and the `.number` within it come on separate lines, so (in English) the result will be exactly as it was before my patch.

The above JSON would result in:

English:
```html
You made <strong><span class="number>234</span> edits</strong> across the project"
```
Norwegian Bokmål:
```html
Du gjorde <strong><span class="number">234</span> redigeringer</strong> på prosjektet"
```
Swahili:
```html
Umefanya <strong>mabadiliko <span class="number">234</span></strong> kwenye mradi mzazi
```


I'd be happy to modify my patch to implement this, so you can see it in action.

jdlrobson commented:
Yeh the strong tag idea seems like a good way to divide this! Nice idea!

I think we will need to be careful with this change though. As an interim step might make sense to do a non-i18n change first to prepare. I can play around with that tomorrow.

I am not too worried about adding plural support. I assume the code in mw.msg is easily adapted.

I will think some more about boxes once the translations come in!

jhsoby committed:
Merge branch 'jdlrobson:main' into main

jhsoby committed:
Adjust how numbers are highlighted

Adjust how the highlight works, and some associated
CSS. This makes it so we still have the i18n advantages,
but there is almost no visible difference from the
previous version.

jhsoby commented:
With the latest commit, I have made everything as close to the original as possible. If you compare them side-by-side, you'll see that there are only a few minor visual differences (a few pixels here and there), which I hope are acceptable.

jdlrobson commented:
Hey I had a look at this today. I wanted to push something out to make progress, so hope you don't mind me stepping on your toes a little.

The main thing I realized we'd need with this approach is sanitization - I don't want to make translatewiki a vector for attack so I added a basic sanitize function.

For now I've retained support for the old style and new style to confirm it's a good solution to the problem before continuing down this path.

Migration of each of the facts looks like this:
https://github.com/jdlrobson/wikipedia-year-in-review/commit/03407be12837622c95981628f599b18f9f5fafff

I would like to chip away at these one at a time to make sure I don't mess up any of these.

Could you pull your badge changes out into a separate patch? If that's too much to ask, I can dissect some changes from your patch myself, with you as author... I know rebases are annoying.

Thanks for helping me think through this!

jhsoby commented:
> Hey I had a look at this today. I wanted to push something out to make progress, so hope you don't mind me stepping on your toes a little.
> 
No worries at all!

> The main thing I realized we'd need with this approach is sanitization - I don't want to make translatewiki a vector for attack so I added a basic sanitize function.
> 
Yeah, I thought about that, but forgot to mention it. Sanitation is definitely good to have, but when the strings come from Translatewiki, they can normally be assumed to be safe – they manually check any HTML tags that look suspicious before exporting stuff (though I'm not sure how that procedure works for GitHub repos). But yeah, better safe than sorry. :-)

> For now I've retained support for the old style and new style to confirm it's a good solution to the problem before continuing down this path.
> 
> Migration of each of the facts looks like this: [03407be](https://github.com/jdlrobson/wikipedia-year-in-review/commit/03407be12837622c95981628f599b18f9f5fafff)
> 
> I would like to chip away at these one at a time to make sure I don't mess up any of these.
> 
> Could you pull your badge changes out into a separate patch? If that's too much to ask, I can dissect some changes from your patch myself, with you as author... I know rebases are annoying.
Sure. By badge changes, do you mean the ones from the share page?
> 
> Thanks for helping me think through this!

jdlrobson commented:
Okay I've updated the messages so we are no longer using LEGO. They still have span and strongs for now, but I am open to revisiting that after giving it exposure to more translators!

jhsoby commented:
> Okay I've updated the messages so we are no longer using LEGO. They still have span and strongs for now, but I am open to revisiting that after giving it exposure to more translators!

Sweet! Thank you very much. Should we close this PR then, perhaps? If I have any other suggestions, I can make separate PRs and/or comments.

jdlrobson commented:
We still need to work out what to do with the facts on the summary screen and plurals!
